asim: simulate store rebalancer

kvoli opened the pull request:
This PR refactors the store rebalancer and adjacent components, without introducing behavioral changes. The aim is to separate and moduralize it's respective parts, for simulaton. A simulator store rebalancer is then added, which uses these refactored parts.

resolves #82631

Release justification: low risk, no behavior changes with the purpose of testing the existing rebalancing system.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/86249)
<!-- Reviewable:end -->

kvoli committed:
kvserver: use relocate options in relocate one

This patch introducs an options struct for relocate one, which is called
iteratively by relocate replicas, for the relocate range command. This
struct contains the necessary external information to determine the next
replica to move, in the chain of replica changes within a relocate
range.

Release note: None

kvoli committed:
kvserver: modularize storerebalancer components

Previously the store rebalancer main function, `rebalanceStore` was
monolithic and intermingled the control loop, algorithm and application
of actions. This patch separates these components into disjoint parts.
This is done to aid in clarity and enable the allocation simulator to
call more production code than was previously possible.

Additionally, changes to adjacent areas have been updated to introduce a
descriptor like structs for allocator input; when unavailable a function
pointer is used instead.

This change is purposed for enabling simulation of the store rebalancer
loop.

No behavior changes were introduced.

Release note: None

kvoli committed:
asim: simulate store rebalancer

This patch includes the store rebalancer within the simulation tick
loop. The store rebalancer simulation logic is maintained in the
`asim/storerebalancer` pkg. Additionally, a higher level operation
controller is introduced, to manage iterative, multi-step or generally
higher level operations such as relocate range. The controller and
operations are maintained in the `asim/op` pkg.

A background tick interval was introduced for components which require
less frequent updates. The components currently included are the state
exchange, workload generators and metrics. The remaining components are
ticked at the interval given; in the simulator speed test this is
lowered to 2s.

resolves #82631

Release note: None

lidorcarmel reviewed (changes requested):
looks great Austen, thanks for your patience, another huge step with this great pr!
my comments are mostly typos and such..

> Reviewed 1 of 44 files at r1, 43 of 43 files at r2, 9 of 9 files at r3, 36 of 36 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86249#-:-NAfPP_7DyOkGIDxK02P:bwb4dht)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @kvoli)

___
*[`pkg/kv/kvserver/replica_rankings.go` line 50 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86249#-N9sA7ds8QCXdHvM4kTh:-N9sA7ds8QCXdHvM4kTi:bnfkthy) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca4a0e2ab90aa56ed1f059f2578d3d43676e2acf/pkg/kv/kvserver/replica_rankings.go#L50)):*
> ```Go
> 	// as the span config for the replica.
> 	DescAndSpanConfig() (*roachpb.RangeDescriptor, roachpb.SpanConfig)
> 	// Desc returns the authoritative range descriptor, acquiring a replica lock in
> ```

nit: perhaps no need to mention the lock here in the interface? some implementations may not have a lock.



_Code quote:_
```Go
	// Desc returns the authoritative range descriptor, acquiring a replica lock in
	// the process.
```
___
*[`pkg/kv/kvserver/store_rebalancer.go` line 159 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86249#-N9sMOMCE-8oFktDYS2U:-N9sMOMCE-8oFktDYS2V:b355lov) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca4a0e2ab90aa56ed1f059f2578d3d43676e2acf/pkg/kv/kvserver/store_rebalancer.go#L159)):*
> ```Go
> // SimulatorStoreRebalancer returns a StoreRebalancer with the given function
> // pointers, for the purposes of breaking dependencies on real code paths.
> func SimulatorStoreRebalancer(
> ```

can we avoid this one? and just pass everything from the caller.
I guess it means we need to pass the allocator, storeID, and the raftStatus func to NewStoreRebalancer, not ideal but maybe better than having 2 constructors, where those can diverge over time?

Anyway I'm not feeling super strongly.. 
___
*[`pkg/kv/kvserver/store_rebalancer.go` line 328 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86249#-N9sPkR2CAm79ImU8Svr:-N9sPkR2CAm79ImU8Svs:bnm8z76) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca4a0e2ab90aa56ed1f059f2578d3d43676e2acf/pkg/kv/kvserver/store_rebalancer.go#L328)):*
> ```Go
> 	// (2) Search for replica rebalances for ranges this store holds a
> 	//     lease for. Where the lowest QPS store that ends up with a voting
> 	//     replica will have have the lease transferred to it.
> ```

double have



_Code quote:_
```Go
have have
```
___
*[`pkg/kv/kvserver/store_rebalancer.go` line 415 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86249#-N9sQnOx8gCyLGb0k4tV:-N9sQnOx8gCyLGb0k4tW:b60qepm) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca4a0e2ab90aa56ed1f059f2578d3d43676e2acf/pkg/kv/kvserver/store_rebalancer.go#L415)):*
> ```Go
> // the outcome RebalanceTargetFound, if there are valid lease transfer targets,
> // NoRebalanceTarget if there are no rebalance targets but the QPS still
> // exceeds the threshold and NoRebalanceTarget if the threshold is not
> ```

s/NoRebalanceTarget/NoRebalanceNeeded/ right?



_Code quote:_
```Go
NoRebalanceTarget
```
___
*[`pkg/kv/kvserver/store_rebalancer.go` line 511 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86249#-N9sSZ8E8UUPxVk8STZh:-N9sSZ8E8UUPxVk8STZi:b-m4nqo1) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca4a0e2ab90aa56ed1f059f2578d3d43676e2acf/pkg/kv/kvserver/store_rebalancer.go#L511)):*
> ```Go
> // true if successful and false if unsuccessful w.r.t the local stores QPS and
> // threshold.
> func (sr *StoreRebalancer) RebalanceStoreOucome(ctx context.Context, rctx *RebalanceContext) bool {
> ```

then why return bool if we're not using it? why not rename this one to something like LogStoreOutcome without the return value?
___
*[`pkg/kv/kvserver/store_rebalancer.go` line 511 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86249#-N9sTPId7UVK1DTYCjlj:-N9sTPId7UVK1DTYCjlk:b-mot2mp) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca4a0e2ab90aa56ed1f059f2578d3d43676e2acf/pkg/kv/kvserver/store_rebalancer.go#L511)):*
> ```Go
> // true if successful and false if unsuccessful w.r.t the local stores QPS and
> // threshold.
> func (sr *StoreRebalancer) RebalanceStoreOucome(ctx context.Context, rctx *RebalanceContext) bool {
> ```

nit: maybe this one should move down right after PostRangeRebalance? looks like it would fit the order you chose better..
___
*[`pkg/kv/kvserver/store_rebalancer.go` line 526 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86249#-N9sSzgF4JibJCpK_42Z:-N9sSzgF4JibJCpK_42_:bwqa437) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca4a0e2ab90aa56ed1f059f2578d3d43676e2acf/pkg/kv/kvserver/store_rebalancer.go#L526)):*
> ```Go
> 	log.Infof(ctx,
> 		"load-based replica transfers successfully brought s%d down to %.2f qps (mean=%.2f, upperThreshold=%.2f)",
> 		rctx.LocalDesc.StoreID, rctx.LocalDesc.Capacity.QueriesPerSecond, rctx.allStoresList.CandidateQueriesPerSecond.Mean, rctx.QPSMaxThreshold)
> ```

is that line too long?
___
*[`pkg/kv/kvserver/store_rebalancer.go` line 533 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86249#-N9sT8Hh08d8SGhTUi9h:-N9sT8Hh08d8SGhTUi9i:bkfzayi) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca4a0e2ab90aa56ed1f059f2578d3d43676e2acf/pkg/kv/kvserver/store_rebalancer.go#L533)):*
> ```Go
> // the outcome RebalanceTargetFound, if there are valid range rebalancing
> // targets, NoRebalanceTarget if there are no rebalance targets but the QPS
> // still exceeds the threshold and NoRebalanceTarget if the threshold is not
> ```

also here, noRebalanceNeeded?



_Code quote:_
```Go
NoRebalanceTarget
```
___
*[`pkg/kv/kvserver/store_rebalancer.go` line 740 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86249#-N9sTlln3huRT6NQnjbm:-N9sTlln3huRT6NQnjbn:b-1mqdab) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca4a0e2ab90aa56ed1f059f2578d3d43676e2acf/pkg/kv/kvserver/store_rebalancer.go#L740)):*
> ```Go
> 	if len(rctx.rebalanceCandidates) == 0 && len(rctx.hottestRanges) >= 0 {
> 		// NB: In practice, the rebalanceCandidates will be populated with
> 		// hottest ranges by the preceeding function call to this. Some
> ```

you mean by `RebalanceLeases()`? maybe be explicit and say it? it wasn't clear to me what the preceeding func call is.



_Code quote:_
```Go
preceeding function call to this
```
___
*[`pkg/kv/kvserver/store_rebalancer_test.go` line 1649 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86249#-N9sVUosDOlDJQCV0Sz3:-N9sVUosDOlDJQCV0Sz4:b-3wde9k) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca4a0e2ab90aa56ed1f059f2578d3d43676e2acf/pkg/kv/kvserver/store_rebalancer_test.go#L1649)):*
> ```Go
> 	status.Lead = uint64(replDesc.ReplicaID)
> 	status.RaftState = raft.StateLeader
> 	status.Commit = 2
> ```

I'm curious, why did you need to change this?
___
*[`pkg/kv/kvserver/asim/asim.go` line 82 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/86249#-N9sX8ke4FPHSDK96hDe:-N9sX8ke4FPHSDK96hDf:bmgxkvo) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9a6bbcd68a68474e85ef2c80eaf41b88c10362a/pkg/kv/kvserver/asim/asim.go#L82)):*
> ```Go
> 	for storeID := range initialState.Stores() {
> 		allocator := initialState.MakeAllocator(storeID)
> 		// TODO(kvoli): Instead of passing in individual settings to contruct
> ```

typo, construct.



_Code quote:_
```Go
contruct
```
___
*[`pkg/kv/kvserver/asim/asim.go` line 84 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/86249#-N9sXNaq2pxSVpJD714s:-N9sXNaq2pxSVpJD714t:b-u2dhrg) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9a6bbcd68a68474e85ef2c80eaf41b88c10362a/pkg/kv/kvserver/asim/asim.go#L84)):*
> ```Go
> 		// TODO(kvoli): Instead of passing in individual settings to contruct
> 		// the per-store ticking controllers, pass a pointer to the simulation
> 		// settings struct - so that settings they may adjusted at runtime.
> ```

rephrase maybe? not sure the last part makes sense to me.
___
*[`pkg/kv/kvserver/asim/op/transfer_lease.go` line 21 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/86249#-NAemSUyESh8P73t2V3B:-NAemSUyESh8P73t2V3C:bb7phvt) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9a6bbcd68a68474e85ef2c80eaf41b88c10362a/pkg/kv/kvserver/asim/op/transfer_lease.go#L21)):*
> ```Go
> )
> 
> // TransferLeaseOp conatins the inforation for a transfer lease operation.
> ```

typo



_Code quote:_
```Go
inforation
```
___
*[`pkg/kv/kvserver/asim/storerebalancer/candidate_replica.go` line 28 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/86249#-NAeuNy99xKlJn-V5OgR:-NAeuNy99xKlJn-V5OgS:b-rghdsz) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9a6bbcd68a68474e85ef2c80eaf41b88c10362a/pkg/kv/kvserver/asim/storerebalancer/candidate_replica.go#L28)):*
> ```Go
> // simulatorReplica is a replica that is being tracked as a potential candidate
> // for rebalancing activities. It maintains a set of methods that enable
> // querying it's state and processing a rebalancing action if taken.
> ```

s/it's/its/ ?



_Code quote:_
```Go
it's
```
___
*[`pkg/kv/kvserver/asim/storerebalancer/candidate_replica.go` line 75 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/86249#-NAeugoy3TEdPDvRMwpx:-NAeugoy3TEdPDvRMwpy:b3zngm6) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9a6bbcd68a68474e85ef2c80eaf41b88c10362a/pkg/kv/kvserver/asim/storerebalancer/candidate_replica.go#L75)):*
> ```Go
> // log.
> func (sr *simulatorReplica) GetFirstIndex() uint64 {
> 	// TODO(kvoli): We always return 1 here as RaftStatus is unimplemented.
> ```

"we always return 1" but we actually return 2 :) I think the comment should say 2 right?



_Code quote:_
```Go
	// TODO(kvoli): We always return 1 here as RaftStatus is unimplemented.
	// When it is implmeneted, this may become variable.
	return 2
```
___
*[`pkg/kv/kvserver/asim/storerebalancer/store_rebalancer.go` line 35 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/86249#-NAex-4FA-49gN3m5Q1q:-NAex-4FA-49gN3m5Q1r:b4i8pwb) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9a6bbcd68a68474e85ef2c80eaf41b88c10362a/pkg/kv/kvserver/asim/storerebalancer/store_rebalancer.go#L35)):*
> ```Go
> 	// rebalance loop nor has any pending operations. It transfers out of this
> 	// state when the sleep interval has completed.
> 	rebalancerSleeping = iota
> ```

should this be:
```
rebalancerSleeping storeRebalancerPhase = iota
```
?



_Code quote:_
```Go
rebalancerSleeping = iota
```
___
*[`pkg/kv/kvserver/asim/storerebalancer/store_rebalancer.go` line 173 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/86249#-NAfGrqEBQs6gSuNBGHF:-NAfGrqEBQs6gSuNBGHG:bh8e9yj) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9a6bbcd68a68474e85ef2c80eaf41b88c10362a/pkg/kv/kvserver/asim/storerebalancer/store_rebalancer.go#L173)):*
> ```Go
> 
> // phaseSleep checks whether the store rebalancer should continue sleeping. If
> // not, it performs a state transfer to prelogue.
> ```

s/prelogue/prologue/



_Code quote:_
```Go
prelogue
```
___
*[`pkg/kv/kvserver/asim/storerebalancer/store_rebalancer.go` line 182 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/86249#-NAfHpqzCjaKe43ShkeC:-NAfHpqzCjaKe43ShkeD:b3902ob) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9a6bbcd68a68474e85ef2c80eaf41b88c10362a/pkg/kv/kvserver/asim/storerebalancer/store_rebalancer.go#L182)):*
> ```Go
> }
> 
> // phasePrologue gathers all the necessary state including hot ranges, store
> ```

nit: maybe worth saying something like "it synchronously transfers into the epilogue or leases phases..." just to say that we should not stay in the prologue... but feel free to ignore.



_Code quote:_
```Go
// pool list and thresholds. It synchronously transfers into the leases phase
// if it passes the should rebalance store check.
```
___
*[`pkg/kv/kvserver/asim/storerebalancer/store_rebalancer.go` line 274 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/86249#-NAfLrls47UHA18DGqyK:-NAfLrls47UHA18DGqyL:b-kqha09) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9a6bbcd68a68474e85ef2c80eaf41b88c10362a/pkg/kv/kvserver/asim/storerebalancer/store_rebalancer.go#L274)):*
> ```Go
> 	// Check whether we should continue to range based rebalancing, after
> 	// trying lease based rebalancing.
> 	if src.sr.TransferToRebalanceRanges(ctx, src.rebalancerState.rctx) {
> ```

this line makes sense to be but wondering - isn't it enough to check whether `outcome` is NoRebalanceNeeded (done here) or NoRebalanceTarget (need to move ranges)?
___
*[`pkg/kv/kvserver/asim/storerebalancer/store_rebalancer.go` line 280 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/86249#-NAfL-gY2Z4XjtKH-yhn:-NAfL-gY2Z4XjtKH-yho:b-grtonn) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9a6bbcd68a68474e85ef2c80eaf41b88c10362a/pkg/kv/kvserver/asim/storerebalancer/store_rebalancer.go#L280)):*
> ```Go
> 	}
> 
> 	// No more rebalancing should be attempted, move to eplogue and clear the
> ```

typo epilogue



_Code quote:_
```Go
eplogue
```
___
*[`pkg/kv/kvserver/asim/storerebalancer/store_rebalancer_test.go` line 56 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/86249#-NAfOXTn-EMAks3rBquv:-NAfOXTn-EMAks3rBquw:b-j6rd9r) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9a6bbcd68a68474e85ef2c80eaf41b88c10362a/pkg/kv/kvserver/asim/storerebalancer/store_rebalancer_test.go#L56)):*
> ```Go
> 	//     transfer r4(1->4),  1500
> 	//     transfer r5(1->5),  1000
> 	//     transfer r6(1->6),  500
> ```

nice test!


<!-- Sent from Reviewable.io -->

kvoli reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86249)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @lidorcarmel)

___
*[`pkg/kv/kvserver/replica_rankings.go` line 50 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86249#-N9sA7ds8QCXdHvM4kTh:-NBn9gwM958YOQl3DFGs:b-dlnt1a) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca4a0e2ab90aa56ed1f059f2578d3d43676e2acf/pkg/kv/kvserver/replica_rankings.go#L50)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

nit: perhaps no need to mention the lock here in the interface? some implementations may not have a lock.
</blockquote></details>

Updated.:
___
*[`pkg/kv/kvserver/store_rebalancer.go` line 159 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86249#-N9sMOMCE-8oFktDYS2U:-NBnAAYNEjCzG081lRzk:b6u8hr9) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca4a0e2ab90aa56ed1f059f2578d3d43676e2acf/pkg/kv/kvserver/store_rebalancer.go#L159)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

can we avoid this one? and just pass everything from the caller.
I guess it means we need to pass the allocator, storeID, and the raftStatus func to NewStoreRebalancer, not ideal but maybe better than having 2 constructors, where those can diverge over time?

Anyway I'm not feeling super strongly.. 
</blockquote></details>

There's a pretty heavy dependence on the replicate queue for the actual constructor. It makes things a lot simpler / less disruptive to keep them separate for now I feel.
___
*[`pkg/kv/kvserver/store_rebalancer.go` line 328 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86249#-N9sPkR2CAm79ImU8Svr:-NBnAWQV3LKd1WMc419z:b-bw8e14) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca4a0e2ab90aa56ed1f059f2578d3d43676e2acf/pkg/kv/kvserver/store_rebalancer.go#L328)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

double have
</blockquote></details>

Updated.
___
*[`pkg/kv/kvserver/store_rebalancer.go` line 415 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86249#-N9sQnOx8gCyLGb0k4tV:-NBnAdzr-LiXgRie4gup:b-vw47bl) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca4a0e2ab90aa56ed1f059f2578d3d43676e2acf/pkg/kv/kvserver/store_rebalancer.go#L415)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

s/NoRebalanceTarget/NoRebalanceNeeded/ right?
</blockquote></details>

Should be NoRebalanceNeeded, updated.
___
*[`pkg/kv/kvserver/store_rebalancer.go` line 511 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86249#-N9sSZ8E8UUPxVk8STZh:-NBnBYPE5nXIsPw7W5dr:b-om63h9) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca4a0e2ab90aa56ed1f059f2578d3d43676e2acf/pkg/kv/kvserver/store_rebalancer.go#L511)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

then why return bool if we're not using it? why not rename this one to something like LogStoreOutcome without the return value?
</blockquote></details>

Renamed - and removed the return value.
___
*[`pkg/kv/kvserver/store_rebalancer.go` line 511 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86249#-N9sTPId7UVK1DTYCjlj:-NBnCY-UE0qso0qm2uEW:b-v12e94) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca4a0e2ab90aa56ed1f059f2578d3d43676e2acf/pkg/kv/kvserver/store_rebalancer.go#L511)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

nit: maybe this one should move down right after PostRangeRebalance? looks like it would fit the order you chose better..
</blockquote></details>

Moved closer, I think they should stay together in this case since they both occur at the end of the range rebalancing phase.
___
*[`pkg/kv/kvserver/store_rebalancer.go` line 526 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86249#-N9sSzgF4JibJCpK_42Z:-NBnCqUJB1BZ8ssga6-v:b-bw8e14) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca4a0e2ab90aa56ed1f059f2578d3d43676e2acf/pkg/kv/kvserver/store_rebalancer.go#L526)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

is that line too long?
</blockquote></details>

Updated.
___
*[`pkg/kv/kvserver/store_rebalancer.go` line 533 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86249#-N9sT8Hh08d8SGhTUi9h:-NBnCwBX8E4TAlqsDMti:b-bw8e14) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca4a0e2ab90aa56ed1f059f2578d3d43676e2acf/pkg/kv/kvserver/store_rebalancer.go#L533)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

also here, noRebalanceNeeded?
</blockquote></details>

Updated.
___
*[`pkg/kv/kvserver/store_rebalancer.go` line 740 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86249#-N9sTlln3huRT6NQnjbm:-NBnD4GhBYbhBGlUexuW:bvdadyw) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca4a0e2ab90aa56ed1f059f2578d3d43676e2acf/pkg/kv/kvserver/store_rebalancer.go#L740)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

you mean by `RebalanceLeases()`? maybe be explicit and say it? it wasn't clear to me what the preceeding func call is.
</blockquote></details>

Updated to be explicit.
___
*[`pkg/kv/kvserver/store_rebalancer_test.go` line 1649 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86249#-N9sVUosDOlDJQCV0Sz3:-NBnDEdmB5D9P67EWM2L:b-7cvmnm) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca4a0e2ab90aa56ed1f059f2578d3d43676e2acf/pkg/kv/kvserver/store_rebalancer_test.go#L1649)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

I'm curious, why did you need to change this?
</blockquote></details>

We were never actually testing the behind raft leader in the store rebalancer correctly. Down the bottom of the no lease transfer to behind replicas, it checks whether the match is >= leader index - 1. Since we had set it to 1, even if it were 0 it would still pass.

So by adding in this, it actually tests the no lease transfer to behind followers.

https://github.com/kvoli/cockroach/blob/b9a6bbcd68a68474e85ef2c80eaf41b88c10362a/pkg/kv/kvserver/store_rebalancer_test.go#L1394-L1398
___
*[`pkg/kv/kvserver/asim/asim.go` line 82 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/86249#-N9sX8ke4FPHSDK96hDe:-NBnEAEV-GKGTK7XQhP-:b-bw8e14) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9a6bbcd68a68474e85ef2c80eaf41b88c10362a/pkg/kv/kvserver/asim/asim.go#L82)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

typo, construct.
</blockquote></details>

Updated.
___
*[`pkg/kv/kvserver/asim/asim.go` line 84 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/86249#-N9sXNaq2pxSVpJD714s:-NBnENbY5J5ReUku5FbE:b-wu4x2t) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9a6bbcd68a68474e85ef2c80eaf41b88c10362a/pkg/kv/kvserver/asim/asim.go#L84)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

rephrase maybe? not sure the last part makes sense to me.
</blockquote></details>

Reworded to be clearer.
___
*[`pkg/kv/kvserver/asim/op/transfer_lease.go` line 21 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/86249#-NAemSUyESh8P73t2V3B:-NBnEU4q3GCx4eBF2Ik0:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9a6bbcd68a68474e85ef2c80eaf41b88c10362a/pkg/kv/kvserver/asim/op/transfer_lease.go#L21)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

typo
</blockquote></details>

Fixed.
___
*[`pkg/kv/kvserver/asim/storerebalancer/candidate_replica.go` line 28 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/86249#-NAeuNy99xKlJn-V5OgR:-NBnEZjU1ThBAhnbxID0:b-qssebl) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9a6bbcd68a68474e85ef2c80eaf41b88c10362a/pkg/kv/kvserver/asim/storerebalancer/candidate_replica.go#L28)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

s/it's/its/ ?
</blockquote></details>

fixed.
___
*[`pkg/kv/kvserver/asim/storerebalancer/candidate_replica.go` line 75 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/86249#-NAeugoy3TEdPDvRMwpx:-NBnEiR_6Ya2jLZ6QKTk:b-ttmkov) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9a6bbcd68a68474e85ef2c80eaf41b88c10362a/pkg/kv/kvserver/asim/storerebalancer/candidate_replica.go#L75)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

"we always return 1" but we actually return 2 :) I think the comment should say 2 right?
</blockquote></details>

That is right, updated.
___
*[`pkg/kv/kvserver/asim/storerebalancer/store_rebalancer.go` line 35 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/86249#-NAex-4FA-49gN3m5Q1q:-NBnExiP7jjjWjLGgpVT:bjmcd6p) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9a6bbcd68a68474e85ef2c80eaf41b88c10362a/pkg/kv/kvserver/asim/storerebalancer/store_rebalancer.go#L35)):*
> rebalancerSleeping storeRebalancerPhase = iota
It should be - updated.
___
*[`pkg/kv/kvserver/asim/storerebalancer/store_rebalancer.go` line 173 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/86249#-NAfGrqEBQs6gSuNBGHF:-NBnF-8rDev6Y1oXrRJ8:b-g0ph17) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9a6bbcd68a68474e85ef2c80eaf41b88c10362a/pkg/kv/kvserver/asim/storerebalancer/store_rebalancer.go#L173)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

s/prelogue/prologue/
</blockquote></details>

The naming here was intentional - like the part before a story.
___
*[`pkg/kv/kvserver/asim/storerebalancer/store_rebalancer.go` line 182 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/86249#-NAfHpqzCjaKe43ShkeC:-NBnFLz79SycJDMLGTN3:b-yf9ny8) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9a6bbcd68a68474e85ef2c80eaf41b88c10362a/pkg/kv/kvserver/asim/storerebalancer/store_rebalancer.go#L182)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

nit: maybe worth saying something like "it synchronously transfers into the epilogue or leases phases..." just to say that we should not stay in the prologue... but feel free to ignore.
</blockquote></details>

Updated wording.
___
*[`pkg/kv/kvserver/asim/storerebalancer/store_rebalancer.go` line 274 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/86249#-NAfLrls47UHA18DGqyK:-NBnFdY4AerqA4NbV_D9:bivzhre) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9a6bbcd68a68474e85ef2c80eaf41b88c10362a/pkg/kv/kvserver/asim/storerebalancer/store_rebalancer.go#L274)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

this line makes sense to be but wondering - isn't it enough to check whether `outcome` is NoRebalanceNeeded (done here) or NoRebalanceTarget (need to move ranges)?
</blockquote></details>

There is added logging within this method call that I wanted to call in both the real code and the simulator.
___
*[`pkg/kv/kvserver/asim/storerebalancer/store_rebalancer.go` line 280 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/86249#-NAfL-gY2Z4XjtKH-yhn:-NBnFipH9ncpWMT1URmx:bj8bld6) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9a6bbcd68a68474e85ef2c80eaf41b88c10362a/pkg/kv/kvserver/asim/storerebalancer/store_rebalancer.go#L280)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

typo epilogue
</blockquote></details>

same as above.
___
*[`pkg/kv/kvserver/asim/storerebalancer/store_rebalancer_test.go` line 56 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/86249#-NAfOXTn-EMAks3rBquv:-NBnFwtJ5fno-4ltAqz_:b-k73z76) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9a6bbcd68a68474e85ef2c80eaf41b88c10362a/pkg/kv/kvserver/asim/storerebalancer/store_rebalancer_test.go#L56)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

nice test!
</blockquote></details>

thanks


<!-- Sent from Reviewable.io -->

lidorcarmel reviewed (approved):
> Reviewed 1 of 46 files at r5, 9 of 9 files at r6, 36 of 36 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86249)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @kvoli)

___
*[`pkg/kv/kvserver/asim/storerebalancer/store_rebalancer.go` line 173 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/86249#-NAfGrqEBQs6gSuNBGHF:-NCWADD-5dD--stR7C3V:bq2q1z1) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9a6bbcd68a68474e85ef2c80eaf41b88c10362a/pkg/kv/kvserver/asim/storerebalancer/store_rebalancer.go#L173)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

The naming here was intentional - like the part before a story.
</blockquote></details>

I'm a bit confused - `prelogue` is not english, right? and also, in the code you call it `phasePrologue` and in the comment it's `prelogue`. not critical of course :) but it does look like a typo..
___
*[`pkg/kv/kvserver/asim/storerebalancer/store_rebalancer.go` line 280 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/86249#-NAfL-gY2Z4XjtKH-yhn:-NCWBZaIDiOwX5uZOO6C:bx11j85) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9a6bbcd68a68474e85ef2c80eaf41b88c10362a/pkg/kv/kvserver/asim/storerebalancer/store_rebalancer.go#L280)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

same as above.
</blockquote></details>

same as above :) the comment has a typo, the code is great.


<!-- Sent from Reviewable.io -->

kvoli reviewed (commented):
TYFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86249#-:-NC_Qpob9TDd3D_kT_8T:bzftgo7)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @lidorcarmel)

___
*[`pkg/kv/kvserver/asim/storerebalancer/store_rebalancer.go` line 173 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/86249#-NAfGrqEBQs6gSuNBGHF:-NC_Qmt_4joxxzcZDfiN:b-v8bge1) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9a6bbcd68a68474e85ef2c80eaf41b88c10362a/pkg/kv/kvserver/asim/storerebalancer/store_rebalancer.go#L173)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

I'm a bit confused - `prelogue` is not english, right? and also, in the code you call it `phasePrologue` and in the comment it's `prelogue`. not critical of course :) but it does look like a typo..
</blockquote></details>

woops my bad - fixed typo.
___
*[`pkg/kv/kvserver/asim/storerebalancer/store_rebalancer.go` line 280 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/86249#-NAfL-gY2Z4XjtKH-yhn:-NC_QoVN8cS1VEa054sB:b-cypxrc) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9a6bbcd68a68474e85ef2c80eaf41b88c10362a/pkg/kv/kvserver/asim/storerebalancer/store_rebalancer.go#L280)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

same as above :) the comment has a typo, the code is great.
</blockquote></details>

updated.


<!-- Sent from Reviewable.io -->

kvoli commented:
bors r=lidorcarmel
